### PR TITLE
feat(Reporting errors to Bugsnag): Update to report all exceptions and HTTP 500s to Bugsnag

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [ ] The description lists any configuration setting(s) that differ from the default settings
- [ ] All tests and CI builds passing

### Description

The OTP middleware has been enhanced to include error reporting to Bugsnag. This covers points where exceptions are caught and HTTP status 500s are reported. No configuration changes are required, however, for this to be enabled a valid BUGSNAG_PROJECT_NOTIFIER_API_KEY must be provided. Details on how to obtain a BUGSNAG_PROJECT_NOTIFIER_API_KEY can be found in the README.

I've included unit tests, but these require a valid BUGSNAG_PROJECT_NOTIFIER_API_KEY in test\env.yml which is not permitted. Because of this, I have disabled the tests so as not to break Travis.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: I find it slightly more readable to have the string message be the first param (something about the longer string literal coming first or maybe because it mimics the Logger signature). Not a blocking change, but I think I would prefer that order.

File: src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java]: Is there a convenient way to log the message inside the `reportErrorToBugsnag` method (ideally without losing the line number/stack trace to come back to)? Just because it's a bit redundant to have the same message repeated. Also, it seems valuable to include tripRequest details here as well (or was there a reason for not doing that?).

File: src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java
Comment (landonreed) [inline src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: Is it possible to create a test that triggers a report somewhere in the actual code (Persistence perhaps)?

**Review by landonreed — CHANGES_REQUESTED** [review]
This looks good, but I wonder if we could add error logging within the reportError method to reduce duplicated messages? Maybe there's not really a great way to do that though...

I would also like to see a test that actually triggers a report within the code itself (rather than just the theoretical tests).

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: What does this sentence mean?

File: src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java]: +1 on the duplicate messages.

File: src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: Would it make sense to create an `/makeerror` endpoint that generates a 500 error so we can catch it?

**Review by binh-dam-ibigroup — COMMENTED** [review]
I concur with @landonreed on his comments, and that we need a way to trigger an error report.

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: Updated.

File: src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java]: This is misleading. It is actually the context, so Bugsnag groups errors by this value. I'm going to refactor as additional information is required.

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: Bugsnag cannot create a report if the exception is null, it causes a NPE. Hence logging this in case it is not provided/null.

File: src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java
Comment (br648) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: @landonreed if @binh-dam-ibigroup suggestion it ok, I will implement.

Comment (landonreed) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: Hmm, perhaps. I don't think we would really want this to be an open HTTP endpoint though (only accessible by Admin users). I'm kind of ambivalent here because this doesn't seem like something we really want to build into the non-test software, but it would be convenient to have a way to verify that things are configured correctly. @evansiroky?

Comment (br648) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: @landonreed - How about creating and documenting a test that attempts to insert a trip request into a stopped Mongo DB instance? No additional non-test code is created and the test is still under the context of a unit test.

Comment (landonreed) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: It will probably fail if you just try to insert a `null` TripRequest, right? That seems like an easy test to write up.

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: Ok. I would change the text to something like "This error is not an exception and will not/cannot be reported to Bugsnag: {...}."

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java]: Updated.

File: src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java
Comment (br648) [reply src/test/java/org/opentripplanner/middleware/BugsnagReportingTest.java]: @landonreed - I've made an update and created a unit test. As you will see this required additional code within the TypedPersistence -> create method and a less conventional unit test. Now that I have done this other methods in the TypedPersistence class could be updated in a similar way, but would like your view on this.

File: src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/otp/OtpRequestProcessor.java]: Updated to be context, message, exception. The message is saved as app info which is basically a key/value pair. It is possible to save a lot more information under app info, but to keep the Bugsnag reporting call to a minimum only a single item can be provided.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: I think this recent refactor of `reportErrorToBugsnag` perhaps should contain some of this logic. Instead of its signature being `String context, String message, Throwable e`, I think it would be more helpful to swap message for a generic `Object badEntity`. I think the easiest approach would be for me to just make a PR to this to demonstrate what I'm thinking. See #44.

**Comment by landonreed** [comment]
Blocked until #44 is merged.

**Review by landonreed — APPROVED** [review]

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java:49]: Is the order for `message` and `context` swapped?

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java:57]: Is the order for `message` and `context` swapped?

**Review by binh-dam-ibigroup — COMMENTED** [review]
Just two questions about the log messages when bugsnag is disabled.

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (landonreed) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java:49]: @br648, when I modified the param names, I forgot to update these log messages. Should we just change the log messages to match the new param names?

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java:49]: yep, updated.

File: src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/bugsnag/BugsnagReporter.java:57]: Updated.

File: src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/persistence/TypedPersistence.java]: Merged #44

**Review by binh-dam-ibigroup — APPROVED** [review]

**Pull request merged**
